src/cpp/session/include/session/SessionProjectTemplate.hpp

```diff
@@ -0,0 +1,261 @@
+   {
+      ProjectTemplateWidgetDescription ptwd;
+      
+      core::Error error = core::json::readObject(
+               object,
+               "parameter", &ptwd.parameter,
+               "type",      &ptwd.type,
+               "label",     &ptwd.label);
+      
+      if (error)
+         LOG_ERROR(error);
+
+      core::json::fillVectorString(
+            object["fields"].get_array(),
+            &(ptwd.fields));
+
+      return ptwd;
+   }
+};
+
+struct ProjectTemplateDescription
+{
+   std::string package;
+   std::string binding;
+   std::string title;
+   std::string subtitle;
+   std::string caption;
+   std::string icon;
+   std::vector<ProjectTemplateWidgetDescription> widgets;
+   
+   static core::Error parseWidget(const std::string& widget, std::string* pWidgetType)
+   {
+      std::string widgetLower = core::string_utils::toLower(widget);
+      for (std::size_t i = 0, n = sizeof(kWidgetTypes); i < n; ++i)
+      {
+         const char* widgetType = kWidgetTypes[i];
+         if (widgetLower == widgetType)
+         {
+            pWidgetType->assign(widgetType);
+            return core::Success();
+         }
+      }
+      
+      return core::systemError(
+               boost::system::errc::protocol_error,
+               ERROR_LOCATION);
+   }
+   
+   static core::Error parseFields(const std::string& fields, std::vector<std::string>* pFields)
+   {
+      std::vector<std::string> parsedFields;
+      core::text::parseCsvLine(
+               fields.begin(),
+               fields.end(),
+               true,
+               &parsedFields);
+      
+      if (parsedFields.empty())
+         return core::systemError(boost::system::errc::protocol_error, ERROR_LOCATION);
+      
+      for (std::size_t i = 0, n = parsedFields.size(); i < n; ++i)
+         parsedFields[i] = core::string_utils::trimWhitespace(parsedFields[i]);
+      
+      *pFields = parsedFields;
+      return core::Success();
+   }
+   
+   template <typename T>
+   static core::Error populate(const core::FilePath& resourcePath,
+                               const T& map,
+                               ProjectTemplateDescription* pDescription)
+   {
+      ProjectTemplateWidgetDescription widget;
+      for (typename T::const_iterator it = map.begin();
+           it != map.end();
+           ++it)
+      {
+         const std::string& key   = it->first;
+         const std::string& value = it->second;
+         
+         // populate primary keys
+         if (key == "Binding")
+            pDescription->binding = value;
+         else if (key == "Title")
+            pDescription->title = value;
+         else if (key == "Subtitle")
+            pDescription->subtitle = value;
+         else if (key == "Caption")
+            pDescription->caption = value;
+         else if (key == "Icon")
+         {
+            // read icon file from disk
+            core::FilePath iconPath = resourcePath.parent().complete(value);
```

Comment:
It might be worth defending against pathologically large files here by skipping icons larger than some reasonable size (we don't want to accidentally try to stream megabytes of base64 data to the client).